### ISSUE TITLE
wierd time cost when training a epoch

### ISSUE THREAD

**Issue by Jordan20160310** [apache/incubator-mxnet#1614]
**As shown below,is the main code of training.**
( i have to add '----' because Space character is always be deleted,please ignore it )
//////////////// 
def _train_multi_device(...)
--------# Now start training
--------for epoch in range(begin_epoch, end_epoch):
--------# Iterate over training data.
(1)-----while True:
(2)--------for data_batch in train_data:
(3)------------executor_manager.load_data_batch(data_batch)
(4)------------executor_manager.forward(is_train=True)
(5)------------executor_manager.backward()
(6)------------if update_on_kvstore:
(7)----------------_update_params_on_kvstore(...)
(8)------------else:
(9)----------------_update_params(...)
(10)----------executor_manager.update_metric(eval_metric, data_batch.label)
(11)----------if epoch_size is not None and nbatch >= epoch_size:
(12)--------------do_reset = False
(13)--------------break
//////////////

**we test according to the following method:**
a)take the first 100 classes of 'imagenet',and make '.rec' data of my own use im2rec.
b)use 'inception-bn' in example to train, and set: 2 gpus, batchsize=64.
c)test in local and distributed mode,and record the time consumption of each operation from (1)to(13).

**what happened below:**
1) in local mode,after training of 'one epoch', time of each operation are sumed as flow:
 total: 1464.029
 load_data_batch            cost=0.393
 forward & backward       cost=2.903
 update_params              cost=122.479
 update_metric                cost=1165.000
 step10 to step3 Intelval  cost=0.774
=====>**QUESTION-1**: 'update_metric' update the 'predict label' by change the value in momory ,why it cost so many time ?

2) in local mode,in order to shorten training time, we don't 'update_metric'(delete it).after training of 'one epoch', time of each operation are sumed as flow:
 total: 1480.09
 load_data_batch             cost=0.378
 forward & backward        cost=2.803
 update_params              cost=121.379
 update_metric                 cost=0.000
 step10 to step3 Intelval   cost=1172.000 
=====>**QUESTION-2**:why the total training time for a epoch between 'update_metric' and 'not update_metric' are scarcely less ?
=====**>QUESTION-3**:why it cost so many time between 'the end of current loop' and 'the start of next loop' (operate (1) and operate (2)) ? what's it doing ?
=====>**QUESTION-4**:how can i accelerate training ?

3）in distributed mode,we get the same result as local mode.
（we set "distributed mode" as：we cut the training list into two parts，and make two rec files，and save this two "rec" files in two worker,when training each worker use his own rec file.）

**Comment by Jordan20160310** [comment]
why the operation "**executor_manager.load_data_batch(data_batch)**"  load  data  batchsize=64  faster  than  data batchsize=32  ? 

batchsize=64   0.35s  needed
batchsize=32   0.69s  needed

**Comment by revilokeb** [comment]
@Jordan20160310 I am looking at a very similar setup, but but have prepared .rec with all 1000 classes. I can roughly confirm your numbers of your question 1. I dont have an answer for you at the moment, but will let you know when I find some as I am finding your observation also strange.

Can I ask you the following questions since I am having trouble to continue learning pre-trained models such as inception-bn (https://github.com/dmlc/mxnet/issues/1439): Are you starting learning from scratch or continue learning pre-trained inception-bn model from model gallery? In case of the latter what accuracy are you getting right after training start? And are you using train_imagenet.py or some scripts of your own?

In my case learning is always starting from scratch although I am providing pre-trained inception model from model gallery when using train_imagenet.py (and also when doing model = mx.model.FeedForward.load(...) and later model.fit(...) after specifying parameters).

**Comment by Jordan20160310** [comment]
@revilokeb  thanks for your answer, here are answer of your questions 
1. learning from scratch.
2. using train_imagenet.py.
3. the training accuracy i get after training start is 0.02 and rise to 0.75 after 7 epoch.  
    --- if you have  accuracy problem here may be the answer: A commonly mistake is forgetting shuffle the image list during packing. This will lead fail of training, eg. ```accuracy``` keeps 0.001 for several rounds.

**Comment by huangjms** [comment]
Any update on this issue? Why the update_metric overhead is so high?

**Comment by guoying1030** [comment]
Any update on this issue? Why the update_metric overhead is so high?
